### PR TITLE
fuzz: Fix misplaced SeedRand::ZEROS

### PR DESCRIPTION
**Pull request by maflcko**
After commit fae63bf13033adec80c7e6d73144a21ea3cfbc6d this must be placed even before test_setup. This is nice, because it makes the usage consistently appear in the first line.

The change is moving a `SeedRandomForTest(SeedRand::ZEROS)` to happen earlier. This is fine, because it will either have no effect, or make the code more deterministic, because after commit fae63bf, no other re-seeding other than `ZEROS` can happen in fuzz tests.

### CODE REVIEW

**Review by marcofleon — APPROVED** [review]
ACK fa91ee10f1c1b9f393451eda5cbcd31609d3f103

**Comment by marcofleon** [comment]
Also, if you're already changing the message, small nit for this line:
```
"Without a solution, fuzz stability and determinism can lead \n"
"to non-reproducible bugs or inefficient fuzzing.\n\n"
```
Do you mean "stability and determinism issues" or something similar (instability and non-determinism)? The sentence is a bit unclear to me, as fuzz stability and determinism is what we're aiming for.

**Comment by marcofleon** [comment]
Re ACK fadd568931a2d21e0f80e1efaf2281f5164fa20e

**Review by brunoerg — APPROVED** [review]
code review ACK fadd568931a2d21e0f80e1efaf2281f5164fa20e

**Comment by ryanofsky** [comment]
> After commit [fae63bf](https://github.com/bitcoin/bitcoin/commit/fae63bf13033adec80c7e6d73144a21ea3cfbc6d) this must be placed even before test_setup.

I think I don't understand what happens if `SeedRandomStateForTest(SeedRand::ZEROS);` is not called before the test setup. Is this causing the fuzz test to just crash on startup? Would be helpful to say what is happening in the PR description.

**Comment by hodlinator** [comment]
I was worried that `g_rng_temp_path` had somehow stopped working again, but I see you needed to replace it with `m_rng` in your diff to make paths deterministic.

It is a dilemma regarding the hash salts, one would like to exercise many values for them, but still get determinism. What would you think about having the fuzz input affect the random seed?

**Review by hodlinator — APPROVED** [review]
ACK fadd568931a2d21e0f80e1efaf2281f5164fa20e

Prior to commit fae63bf13033adec80c7e6d73144a21ea3cfbc6d (#31486):
* `utxo_total_supply` fuzz target was initializing `ChainTestingSetup` which would call `SeedRandomStateForTest(SeedRand::FIXED_SEED)`.
* `SeedRandomStateForTest(SeedRand::FIXED_SEED)` would seed randomness using a value from a static variable initialized once during process startup.
* This was making each fuzz target execution (input) use the same value.

After fae63bf13033adec80c7e6d73144a21ea3cfbc6d:
* Due to the added `if constexpr (!G_FUZZING)`-guard inside of the `ChainTestingSetup`-initialization, we were no longer resetting the PRNG for every fuzz input (aside from the explicit `SeedRandomStateForTest(SeedRand::ZEROS)`-calls in the fuzz targets).
* This lead to PRNG state leakage from the previous fuzz input into the next input up until `utxo_total_supply` itself explicitly called `SeedRandomStateForTest(SeedRand::ZEROS)`.
* This PR (commit fadd568931a2d21e0f80e1efaf2281f5164fa20e) fixes that leakage by calling `SeedRandomStateForTest(SeedRand::ZEROS)` before `ChainTestingSetup`-initialization.

### Testing

Added this to the `utxo_total_supply` target:
```C++
std::ofstream{"/tmp/fuzz_data", std::ios_base::app} << HexStr(FastRandomContext().rand256()) << std::endl;
```
When placed before `SeedRandomStateForTest(SeedRand::ZEROS)`, `/tmp/fuzz_data` would contain this for 4 runs:
```
b0fd14ff96a0bda154c329082c9c6533bb4c9473bf5dde138f82c9ac5553d958
daaa4b425bfa3725e2b0f59f0fabf4ee3387123c569425cbcf8808296acc6f4e
daaa4b425bfa3725e2b0f59f0fabf4ee3387123c569425cbcf8808296acc6f4e
daaa4b425bfa3725e2b0f59f0fabf4ee3387123c569425cbcf8808296acc6f4e
```
\- Meaning PRNG state would leak from the previous run, but it would be the same leaked state on the next run. This is not guaranteed if the fuzz target is complex enough though, as different fuzz inputs could lead to different amount of randomness being used.

When placed after `SeedRandomStateForTest(SeedRand::ZEROS)`, one gets the same on each run:
```
b0fd14ff96a0bda154c329082c9c6533bb4c9473bf5dde138f82c9ac5553d958
b0fd14ff96a0bda154c329082c9c6533bb4c9473bf5dde138f82c9ac5553d958
b0fd14ff96a0bda154c329082c9c6533bb4c9473bf5dde138f82c9ac5553d958
b0fd14ff96a0bda154c329082c9c6533bb4c9473bf5dde138f82c9ac5553d958
```

---

### Potential follow-up

Conceptually I think it would be an improvement to seed randomness using fuzz input instead of `ZEROS`, but that should be done in another PR. It would probably lead to marginally more test coverage while still retaining determinism.

The initialization-step would still use `ZEROS` I guess since it has no given fuzz input-data AFAIK.

File: src/test/fuzz/util/check_globals.cpp [hunk 1]
```diff
@@ -24,12 +24,12 @@ struct CheckGlobalsImpl {
                          "The current fuzz target used the global random state.\n\n"
 
                          "This is acceptable, but requires the fuzz target to call \n"
-                         "SeedRandomStateForTest(SeedRand::ZEROS) at the beginning \n"
-                         "of processing the fuzz input.\n\n"
+                         "SeedRandomStateForTest(SeedRand::ZEROS) in the first line \n"
+                         "of the FUZZ_TARGET function.\n\n"
```
[the remark this answers is not in the document]
Comment (hodlinator) [reply src/test/fuzz/util/check_globals.cpp:28]: Was thinking there may creep up something else that also needs initializing really early and either is unaffected by randomness or somehow needs to affect randomness seeding (yes, contrived).

Then there is the idea to have a mixin seeding the randomness internally, which would change this part regardless.

**Comment by ryanofsky** [comment]
Thanks for the explanations. The reason I was confused earlier is that I mistakenly thought `CheckGlobalsImpl` would crash if a fuzz test ever used the global RNG without seeding it with zeroes first.

But (unfortunately?) the checker doesn't really check that. It only checks that if a fuzz test uses the RNG, then it must seed it with zeros at some point during the test. The checker doesn't check that seeding happens **before** using the RNG.

So this PR does fix some nondeterminism, but the nondeterminism was undetected by the checker, so it was not causing any crash.

**Review by Daisha31 — APPROVED** [review]

**Review by Daisha31 — APPROVED** [review]

**Review by Daisha31 — APPROVED** [review]

**Review by Daisha31 — APPROVED** [review]

**Review by Daisha31 — APPROVED** [review]

**Review by Daisha31 — APPROVED** [review]

**Review by Daisha31 — APPROVED** [review]

**Pull request merged**